src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -177,23 +177,45 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
     }
 
     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Fuction redirect to random page
+        // Fuction redirect to random page of the same type
         // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
         Long mealId = Long.parseLong(getParameter(request, "id", "0"));
+        String mealType = null;
 
         Query query = new Query("Meal");
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
+
+        for (Entity entity : results.asIterable()) {
+            Long id = (Long)entity.getProperty("id");
+            if (id != null && id.equals(mealId)) {
+                mealType = (String)entity.getProperty("type");
+            }
+        }
+
+        ArrayList<Long> sameTypeIdList = new ArrayList<>();
         ArrayList<Long> idList = new ArrayList<>();
         for (Entity entity : results.asIterable()) {
             Long id = (Long)entity.getProperty("id");
+            String type = (String)entity.getProperty("type");
             if (id != null && !id.equals(mealId)) {
                 idList.add(id);
             }
+            if (id != null && !id.equals(mealId) && type != null && type.equals(mealType)) {
+                sameTypeIdList.add(id);
+            }
```

Comment:
This filtering can also be done via query filters:
Make a separate query that has a filter on "type" to be equal to `mealType` and a filter on "id" to be NOT equal to `mealId`.
Offloading all these operations onto the database instead of the application is preffered, because database can do this efficiently and there's less data to be streamed and held by the servlet itself.